src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java

```diff
@@ -0,0 +1,132 @@
+import static hudson.util.TimeUnit2.MINUTES;
+
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import hudson.Extension;
+import hudson.model.AbstractBuild;
+import hudson.model.Cause;
+import hudson.model.BuildListener;
+import hudson.model.Executor;
+import hudson.model.Result;
+import hudson.plugins.build_timeout.BuildTimeOutOperation;
+import hudson.plugins.build_timeout.BuildTimeOutOperationDescriptor;
+
+/**
+ * Abort the build.
+ */
+public class AbortAndRestartOperation extends BuildTimeOutOperation {
+    private final int maxRestarts;
+    
+    /**
+     * @return max restarts.
+     */
+    public int getMaxRestarts() {
+        return maxRestarts;
+    }
+    
+    @DataBoundConstructor
+    public AbortAndRestartOperation(int maxRestarts){
+        this.maxRestarts = maxRestarts;
+    }
+       
+    /**
+     * @param build
+     * @param listener
+     * @param effectiveTimeout
+     * @return
+     * @see hudson.plugins.build_timeout.BuildTimeOutOperation#perform(hudson.model.AbstractBuild, hudson.model.BuildListener, long)
+     */
+    @Override
+    public boolean perform(AbstractBuild<?, ?> build, BuildListener listener, long effectiveTimeout) {
+        long effectiveTimeoutMinutes = MINUTES.convert(effectiveTimeout,MILLISECONDS);
+        // Use messages in hudson.plugins.build_timeout.Messages for historical reason.
+        listener.getLogger().println(hudson.plugins.build_timeout.Messages.Timeout_Message(
+                effectiveTimeoutMinutes,
+                hudson.plugins.build_timeout.Messages.Timeout_Aborted())
+        );
+        
+        Executor e = build.getExecutor();
+        if (e != null) {
+            e.interrupt(Result.ABORTED);
+        }
+        
+        if (checkMaxRestarts(build)) {
+            build.getRootBuild().getProject().scheduleBuild(new BuildTimeoutAbortAndRestartCause());
+        }
+        return true;
+    }
+    
+    public class BuildTimeoutAbortAndRestartCause extends Cause {
+        
+
+        /**
+         * Constructor.
+         * 
+         * @param s
+         *            The reason/cause for restart.
+         */
+        public BuildTimeoutAbortAndRestartCause() {
+            super();
+            
+        }
+
+        @Override
+        public String getShortDescription() {
+            return "Build Timeout - Abort and Restart";
+        
+        }
+    }
+    
+    private boolean checkMaxRestarts(AbstractBuild<?, ?> build) {
+        if (this.maxRestarts <= 0) {
+            return true;
+        }
+        int count = 0;
+
+        // count the number of restarts for the current project
+        while (build != null && build.getCause(BuildTimeoutAbortAndRestartCause.class) != null) {
+            
+            count++;
+         
+            if (count >= this.maxRestarts) {
+                return false;
+            }
+            build = build.getPreviousBuild();
+        }
+        return true;
+    }

```

Comment:
> Do you mean to write retrying counts in Cause of each build?

Or to link to the build that triggered the current build.

> And why exactly is the code above less stable?

`getPreviousBuild` doesn't always return the last retrying build.
Users can schedule builds while another build is running. That can result wrong retrying counts.